### PR TITLE
Allow local date for inline exclude newer

### LINKED ISSUE #6555: An exclude newer extension for PEP 723
This would be a one liner you could add that to make most scripts 99% reproducible using uv's awesome `--exclude-newer`

### PR DESCRIPTION
**Pull request by hauntsaninja**
Resolves #6555

### CODE REVIEW

**Review by zanieb — APPROVED** [review]
Thank you!

File: crates/uv-resolver/src/exclude_newer.rs [hunk 1]
```diff
@@ -59,6 +59,16 @@ impl FromStr for ExcludeNewer {
...
 
+impl<'de> serde::Deserialize<'de> for ExcludeNewer {
+    fn deserialize<D>(deserializer: D) -> Result<ExcludeNewer, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let s = String::deserialize(deserializer)?;
+        FromStr::from_str(&s).map_err(serde::de::Error::custom)
+    }
+}
```
Comment (zanieb) [inline crates/uv-resolver/src/exclude_newer.rs:70]: cc @BurntSushi I remember some recent discussion about local dates?

**Comment by zanieb** [comment]
@hauntsaninja on further thought, isn't it _wrong_ to allow a local date for `exclude-newer` in files that are intended to be version controlled and distributed? It seems like a bit of a correctness problem to have the script resolve to different dependencies depending on who is running it. It might make sense to require people to use a proper timestamp here that will be correct regardless of the execution timezone.

I think this notably differs from `UV_EXCLUDE_NEWER` which is configured on a single machine and not easily distributed.

**Comment by hauntsaninja** [comment]
The use case I'm thinking of is "I wrote a script a few years ago and I want it to mostly run the same, maximising convenience". My guess is this use case will be much more common than inter-timezone distribution. We could also resolve different dependencies for different people depending on arch or python version as well, it's why I phrased this as "improving" reproducibility.

**Comment by zanieb** [comment]
I feel like if we want to maximize convenience we should allow emitting an exclude newer pin or, even better, we should be able to write a lock to scripts. Hm.

If you're willing to separate the change from the docs, we can merge the docs right away. I think we'll wait for @BurntSushi to be back from vacation to make a decision on the local dates.

**Review by BurntSushi — COMMENTED** [review]
I think this is making the `serde::Deserialize` trait implementation on `ExcludeNewer` consistent with its `FromStr` trait implementation.

I am still pretty iffy on supporting local dates here though. When I added support for it to `ExcludeNewer`, I was thinking about its use on the command line or in user-specific configuration files. But this seems to be expanding its scope to a point where I think it actually becomes problematic. If, for example, `exclude-newer = "2023-10-16"` is put into a script and then shared with someone else, that date may be interpreted as an entirely different instant than is intended. It might be better to just keep it restricted to RFC 3339 timestamps. Although, the difference in the `Deserialize` and `FromStr` trait implementation seems like a footgun.

**Comment by BurntSushi** [comment]
Related: https://github.com/astral-sh/uv/issues/6820

**Comment by hauntsaninja** [comment]
Actually one possibly silly thought, should we make the short form just be a UTC timestamp? In the short form I think it's arguably already a little unclear what the boundary semantics are (e.g. `--exclude-newer 2024-08-28` does kind of read like `2024-08-28T23:59:59` to me).

This is also nice if you have a cloud machine on UTC time and you want to reproduce a command using the short form on your local machine. It's technically a breaking change, but it seems maybe saner, and I have a hard time thinking about the user profile who would be broken by it.

**Comment by hauntsaninja** [comment]
Anyway, sounds like I should split the PRs regardless

**Comment by BurntSushi** [comment]
> Actually one possibly silly thought, should we make the short form just be a UTC timestamp?

This is what it was originally, but I changed it. It's generally bad juju to just assume UTC for an unadorned date. I think we either don't support an unadorned date or we assume it's in the user's local time zone. But in this case, "user's local time zone" creates a configuration that isn't self-contained.

In terms of the time for a date without a time, it's common practice to interpret it as the first instant on that date (which, interestingly, [may not be midnight](https://docs.rs/jiff/latest/jiff/struct.Zoned.html#method.start_of_day).

**Comment by hauntsaninja** [comment]
Sounds good!

**Pull request closed without merging**